modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -207,17 +215,25 @@ public void deactivate() {
      * @return Future completes true when the lease will not prolong in the future, false otherwise.
      */
     public CompletableFuture<Boolean> denyLease(ReplicationGroupId grpId, Lease lease) {
-        var leaseKey = ByteArray.fromString(PLACEMENTDRIVER_PREFIX + grpId);
-
-        byte[] leaseRaw = lease.bytes();
-
         Lease deniedLease = lease.denyLease();
 
         leaseNegotiator.onLeaseRemoved(grpId);
 
+        Collection<Lease> leases = leaseTracker.leasesCurrent();
+        List<Lease> renewedLeases = new ArrayList<>();
+        for (Lease ls : leases) {
+            if (ls.replicationGroupId().equals(grpId)) {
+                renewedLeases.add(deniedLease);
+            } else {
+                renewedLeases.add(ls);
+            }
+        }
+
+        var key = PLACEMENTDRIVER_LEASES_KEY;
+
         return msManager.invoke(
-                value(leaseKey).eq(leaseRaw),
-                put(leaseKey, deniedLease.bytes()),
+                or(notExists(key), value(key).eq(new LeaseBatch(leaseTracker.leasesCurrent()).bytes())),
+                put(key, new LeaseBatch(renewedLeases).bytes()),
```

Comment:
Here you serialize two times, each time when .bytes() is invoked.
Do this only once.

Reply:
here different collections are serialized